tests/test_pre_shape.py

```diff
@@ -0,0 +1,239 @@
+
+        result = self.space.is_tangent(vector, point)
+        self.assertTrue(~ result)
+
+        tangent_vec = self.space.to_tangent(vector, point)
+        result = self.space.is_tangent(tangent_vec, point)
+        self.assertTrue(result)
+
+        vec = gs.array([tangent_vec, vector])
+        result = self.space.is_tangent(vec, point)
+        expected = gs.array([True, False])
+        self.assertAllClose(result, expected)
+
+    @geomstats.tests.np_and_pytorch_only
+    def test_vertical_projection(self):
+        w = gs.random.rand(self.m_ambient, self.k_landmarks)
+        point = self.space.random_uniform()
+        tan = self.space.to_tangent(w, point)
+        vertical = self.space.vertical_projection(tan, point)
+        transposed_point = Matrices.transpose(point)
+
+        tmp_expected = gs.matmul(tan, transposed_point)
+        expected = tmp_expected - Matrices.transpose(tmp_expected)
+
+        tmp_result = gs.matmul(vertical, transposed_point)
+        result = tmp_result - Matrices.transpose(tmp_result)
+        self.assertAllClose(result, expected)
+
+    @geomstats.tests.np_and_pytorch_only
+    def test_vertical_projection_vectorization(self):
+        w = gs.random.rand(self.n_samples, self.m_ambient, self.k_landmarks)
+        point = self.space.random_uniform(self.n_samples)
+        tan = self.space.to_tangent(w, point)
+        vertical = self.space.vertical_projection(tan, point)
+        transposed_point = Matrices.transpose(point)
+
+        tmp_expected = gs.matmul(tan, transposed_point)
+        expected = tmp_expected - Matrices.transpose(tmp_expected)
+
+        tmp_result = gs.matmul(vertical, transposed_point)
+        result = tmp_result - Matrices.transpose(tmp_result)
+        self.assertAllClose(result, expected)
+
+    @geomstats.tests.np_and_pytorch_only
+    def test_horizontal_projection(self):
+        w = gs.random.rand(self.m_ambient, self.k_landmarks)
+        point = self.space.random_uniform()
+        tan = self.space.to_tangent(w, point)
+        horizontal = self.space.horizontal_projection(tan, point)
+        transposed_point = Matrices.transpose(point)
+        result = gs.matmul(horizontal, transposed_point)
+        expected = Matrices.transpose(result)
+
+        self.assertAllClose(result, expected)
+
+    @geomstats.tests.np_and_pytorch_only
+    def test_horizontal_projection_vectorized(self):
+        w = gs.random.rand(self.n_samples, self.m_ambient, self.k_landmarks)
+        point = self.space.random_uniform(self.n_samples)
+        tan = self.space.to_tangent(w, point)
+        horizontal = self.space.horizontal_projection(tan, point)
+        transposed_point = Matrices.transpose(point)
+        result = gs.matmul(horizontal, transposed_point)
+        expected = Matrices.transpose(result)
+
+        self.assertAllClose(result, expected)
+
+    @geomstats.tests.np_and_pytorch_only
+    def test_horizontal_and_is_horizontal(self):
```

Comment:
`test_horizontal_and_is_tangent`
-> let's use the name of the function being tested, here it is `is_tangent`